### PR TITLE
Kafka loadtest modifications.

### PR DESCRIPTION
**Pull request by aechalf-g**
Kafka topics are now automatically deleted and recreated upon running a loadtest; this ensures no backlog in topics causing performance issues. 
Address of zookeeper hosts will now be a required flag when running a test with Kafka brokers.

### CODE REVIEW

File: load-test-framework/src/main/java/com/google/pubsub/flic/Driver.java [hunk 1]
```diff
@@ -193,6 +193,13 @@
   private String broker;
 
   @Parameter(
+      names = {"--zookeeper_ip_address"},
+      description = "The network addresses of the Zookeeper clients, comma separated."
+  )
+
```
Comment (mdietz94) [inline load-test-framework/src/main/java/com/google/pubsub/flic/Driver.java]: Remove whitespace

File: load-test-framework/src/main/java/com/google/pubsub/flic/Driver.java [hunk 2]
```diff
@@ -372,6 +379,11 @@ public void run(BiFunction<String, Map<ClientParams, Integer>, Controller> contr
           broker != null || (kafkaPublisherCount == 0 && kafkaSubscriberCount == 0),
           "If using Kafka you must provide the network address of your broker using the"
               + "--broker flag.");
+      Preconditions.checkArgument(
```
Comment (mdietz94) [inline load-test-framework/src/main/java/com/google/pubsub/flic/Driver.java]: Do we no longer want to support pure Kafka without Zookeeper? I think Zookeeper should be optional.

File: load-test-framework/src/main/java/com/google/pubsub/flic/controllers/GCEController.java [hunk 3]
```diff
@@ -70,6 +77,8 @@
       "projects/ubuntu-os-cloud/global/images/ubuntu-1604-xenial-v20160930"; // Ubuntu 16.04 LTS
   private static final int ALREADY_EXISTS = 409;
   private static final int NOT_FOUND = 404;
+  private static final int REPLICATION_FACTOR = 2;
+  private static final int PARTITIONS = 100;
```
Comment (mdietz94) [inline load-test-framework/src/main/java/com/google/pubsub/flic/controllers/GCEController.java]: These should be flags, not constants.

File: load-test-framework/src/main/java/com/google/pubsub/flic/output/SheetsService.java [hunk 4]
```diff
@@ -124,6 +124,7 @@ public void sendToSheets(String sheetId, Map<ClientType, Controller.LoadtestStat
           new ValueRange().setValues(values.get(1))).setValueInputOption("USER_ENTERED").execute();
     } catch (IOException e) {
       log.error("Error publishing to spreadsheet: " + sheetId);
+      log.error("Here's the exception: " + e);
```
Comment (mdietz94) [inline load-test-framework/src/main/java/com/google/pubsub/flic/output/SheetsService.java]: Put this exception in the above message.

File: load-test-framework/src/main/java/com/google/pubsub/flic/output/SheetsService.java [hunk 5]
```diff
@@ -188,9 +189,9 @@ public void sendToSheets(String sheetId, Map<ClientType, Controller.LoadtestStat
       valueRow.add(new DecimalFormat("#.##").format(
           (double) LongStream.of(
               stats.bucketValues).sum() / stats.runningSeconds * Client.messageSize / 1000000.0));
-      valueRow.add(LatencyDistribution.getNthPercentile(stats.bucketValues, 95.0));
-      valueRow.add(LatencyDistribution.getNthPercentile(stats.bucketValues, 99.0));
-      valueRow.add(LatencyDistribution.getNthPercentile(stats.bucketValues, 99.9));
+      valueRow.add(LatencyDistribution.getNthPercentileMidpoint(stats.bucketValues, 50.0));   // changed this from 95 to 50 to mirror John's tests
```
Comment (mdietz94) [inline load-test-framework/src/main/java/com/google/pubsub/flic/output/SheetsService.java]: Remove this comment.

File: load-test-framework/src/main/proto/loadtest.proto [hunk 6]
```diff
@@ -85,6 +85,9 @@ message KafkaOptions {
 
   // The length of time to poll for.
   google.protobuf.Duration poll_duration = 2;
+
+  // The network address(es) of the Zookeeper hosts.
```
Comment (mdietz94) [inline load-test-framework/src/main/proto/loadtest.proto]: host(s).

File: load-test-framework/src/main/java/com/google/pubsub/flic/Driver.java (hunk 2 above)
Comment (aechalf-g) [inline load-test-framework/src/main/java/com/google/pubsub/flic/Driver.java]: This depends - if no Zookeeper is provided, then the topic deletion/creation will be optional.

File: load-test-framework/src/main/java/com/google/pubsub/flic/controllers/GCEController.java [hunk 7]
```diff
@@ -146,49 +144,58 @@ private GCEController(String projectName, Map<String, Map<ClientParams, Integer>
...
-              try {
-                AdminUtils.deleteTopic(zookeeperUtils, topic);
-              } catch (ZkNodeExistsException e) {
-                log.info("Topic " + topic + " already marked for delete");
-                kafkaFuture.setException(e);
-                return;
+    // If the Zookeeper host is provided, delete and recreate the topic
+    // in order to eliminate performance issues from backlogs.
+    if (!Client.zookeeperIpAddress.isEmpty()) {
+      log.info("ATTENTION WE R IN TOPIC STUFF FOR KAFKA");
```
Comment (mdietz94) [inline load-test-framework/src/main/java/com/google/pubsub/flic/controllers/GCEController.java]: remove

File: load-test-framework/src/main/java/com/google/pubsub/flic/controllers/GCEController.java [hunk 8]
```diff
@@ -146,49 +144,58 @@ private GCEController(String projectName, Map<String, Map<ClientParams, Integer>
...
+                } catch (ZkNodeExistsException e) {
+                  log.info("Topic " + topic + " already marked for delete");
+                  kafkaFuture.setException(e);
+                  return;
+                }
+              } else {
+                log.info("Topic " + topic + " does not exist, no need to delete");
+              }
+              while (AdminUtils.topicExists(zookeeperUtils, topic)) {
+                // waiting for topic to delete before recreating
```
Comment (mdietz94) [inline load-test-framework/src/main/java/com/google/pubsub/flic/controllers/GCEController.java]: Rephrase.

File: load-test-framework/src/main/java/com/google/pubsub/flic/controllers/GCEController.java [hunk 9]
```diff
@@ -225,8 +232,8 @@ private GCEController(String projectName, Map<String, Map<ClientParams, Integer>
       // Wait for files and instance groups to be created.
       Futures.allAsList(pubsubFutures).get();
       log.info("Pub/Sub actions completed.");
-      Futures.allAsList(kafkaFutures).get();
-      log.info("Kafka actions completed.");
+      //Futures.allAsList(kafkaFutures).get();
```
Comment (mdietz94) [inline load-test-framework/src/main/java/com/google/pubsub/flic/controllers/GCEController.java]: These comments seem like a mistake.

File: load-test-framework/src/main/java/com/google/pubsub/flic/output/SheetsService.java [hunk 10]
```diff
@@ -124,7 +124,6 @@ public void sendToSheets(String sheetId, Map<ClientType, Controller.LoadtestStat
           new ValueRange().setValues(values.get(1))).setValueInputOption("USER_ENTERED").execute();
     } catch (IOException e) {
       log.error("Error publishing to spreadsheet: " + sheetId);
```
Comment (mdietz94) [inline load-test-framework/src/main/java/com/google/pubsub/flic/output/SheetsService.java]: You can add the exception here.

**Comment by denyska** [comment]
@mdietz94 do you mind to rebase branch against master, travis.yml was recently changed and now passes on master.

**Comment by mdietz94** [comment]
I don't mind merge vs. rebase, I'm glad travis.yml was fixed. @aechalf-g owns this PR

**Pull request merged**